### PR TITLE
Avar2

### PR DESCRIPTION
**Pull request by behdad**

### CODE REVIEW

**Comment by anthrotype** [comment]
what's the status of this?

**Comment by behdad** [comment]
> what's the status of this?

Needs rework. I ported avar to otData, which breaks the old API. Need to add a compatibility layer in between. Also to figure out how to expose the avar2 portion.

**Comment by behdad** [comment]
This PR so far moves avar parsing to otData and keeps previous API. No handcoded API is added for the avar2 part and that part can be accessed through the .table mechanism. Since that involves a variationStore, I feel like this is enough of an abstraction. WDYT @anthrotype?

File: Tests/ttLib/tables/_a_v_a_r_test.py [hunk 1]
```diff
@@ -47,9 +47,7 @@ def test_decompile(self):
     def test_decompile_unsupportedVersion(self):
         avar = table__a_v_a_r()
         font = self.makeFont(["wdth", "wght"])
-        self.assertRaises(
-            TTLibError, avar.decompile, deHexStr("02 01 03 06 00 00 00 00"), font
-        )
+        TTLibError, avar.decompile, deHexStr("02 01 03 06 00 00 00 00"), font
```
Comment (anthrotype) [inline Tests/ttLib/tables/_a_v_a_r_test.py:50]: this line looks incomplete, where's the assert gone?

File: Lib/fontTools/ttLib/tables/_a_v_a_r.py [hunk 2]
```diff
@@ -54,44 +45,41 @@ class table__a_v_a_r(DefaultTable.DefaultTable):
...
-            numPairs = struct.unpack(">H", data[pos : pos + 2])[0]
-            pos = pos + 2
-            for _ in range(numPairs):
-                fromValue, toValue = struct.unpack(">hh", data[pos : pos + 4])
-                segments[fi2fl(fromValue, 14)] = fi2fl(toValue, 14)
-                pos = pos + 4
+            for segment in segmentMap.AxisValueMap:
+                segments[segment.FromCoordinate] = segment.ToCoordinate
 
     def toXML(self, writer, ttFont):
```
[the remark this answers is not in the document]
Comment (behdad) [reply Lib/fontTools/ttLib/tables/_a_v_a_r.py:84]: You are right I believe.

File: Tests/ttLib/tables/_a_v_a_r_test.py (hunk 1 above)
Comment (anthrotype) [reply Tests/ttLib/tables/_a_v_a_r_test.py:50]: they we should just remove the whole test I guess

Comment (behdad) [reply Tests/ttLib/tables/_a_v_a_r_test.py]: Just did that.

File: Lib/fontTools/ttLib/tables/_a_v_a_r.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (behdad) [reply Lib/fontTools/ttLib/tables/_a_v_a_r.py:87]: Okay, XML seems to work now.

File: Lib/fontTools/ttLib/tables/otTables.py [hunk 3]
```diff
@@ -711,8 +711,9 @@ def preWrite(self, font):
         if mapping is None:
             mapping = self.mapping = {}
 
-        glyphOrder = font.getGlyphOrder()
-        mapping = [mapping[g] for g in glyphOrder]
+        if type(mapping) == dict:
```
Comment (anthrotype) [inline Lib/fontTools/ttLib/tables/otTables.py:714]: why did you change this? When it is _not_ a dict? And what about the missing `else` when it is not `type(mapping) == dict`?

I believe that you don't need to change `VarIdxMap` class, since you aren't even using it here, you're using `DeltaSetIndexMap` class which doesn't use GIDs internally and has a list of integers as `self.mapping`

Comment (behdad) [reply Lib/fontTools/ttLib/tables/otTables.py:714]: Yeah I changed this before I started using `DeltaSetIndexMap`.  I'll revert these changes.

**Comment by behdad** [comment]
I added a test. It doesn't test `fromXML` or check what `toXML` produces but tests the rest.

**Comment by behdad** [comment]
Added fromXML as well.

File: Tests/ttLib/tables/_a_v_a_r_test.py [hunk 4]
```diff
@@ -91,14 +88,92 @@ def makeFont(axisTags):
...
+            {},
+            {"wdth": -0.8},
+        ]
+
+        model = models.VariationModel(master_locations_normalized, axisTags)
+        store_builder = varStore.OnlineVarStoreBuilder(axisTags)
+        store_builder.setModel(model)
+        varIdxes = {}
+        for axis in axisTags:
+            masters = [m.get(axis, 0) * (1 << 14) for m in data]
```
Comment (anthrotype) [inline Tests/ttLib/tables/_a_v_a_r_test.py:125]: `floatToFixed(v, 14)` ?

File: Tests/ttLib/tables/_a_v_a_r_test.py [hunk 5]
```diff
@@ -91,14 +88,92 @@ def makeFont(axisTags):
...
+
+        xml1 = BytesIO()
+        writer = XMLWriter(xml1)
+        font["avar"].toXML(writer, font)
+
+        xml2 = BytesIO()
+        writer = XMLWriter(xml2)
+        font2["avar"].toXML(writer, font2)
+
+        # We can't compare because VarStore goes through transformation during compile :(
```
[the remark this answers is not in the document]
Comment (behdad) [reply Tests/ttLib/tables/_a_v_a_r_test.py:167]: It reorders the columns to minimize wordCount for example.

[the remark this answers is not in the document]
Comment (behdad) [reply Tests/ttLib/tables/_a_v_a_r_test.py:167]: Also seems to add some Format attributes.

[the remark this answers is not in the document]
Comment (behdad) [reply Tests/ttLib/tables/_a_v_a_r_test.py:167]: Also VarIdxMap seems to get a Format argument during compile.

[the remark this answers is not in the document]
Comment (behdad) [reply Tests/ttLib/tables/_a_v_a_r_test.py:167]: > It reorders the columns to minimize wordCount for example.

/me checks if that's during compile or optimize()

[the remark this answers is not in the document]
Comment (behdad) [reply Tests/ttLib/tables/_a_v_a_r_test.py:167]: Huh. Looks like if I add Format to VarIdxMap, it passes the comparison.

I actually think that Format should not exist on the VarIdxMap... I'll file a separate issue for that.

**Review by anthrotype — APPROVED** [review]
